Use correct HTTP status codes for error responses

toupeira opened the pull request:
### Summary

Fixes https://github.com/doorkeeper-gem/doorkeeper/issues/550

Changes the HTTP status codes for error responses to conform with the OAuth specs, as discussed in the issue.

https://tools.ietf.org/html/rfc6749#section-5.2:

- a 400 status should be returned for most error conditions
  - implemented
- a 401 MAY be be returned for an `invalid_client` error, and MUST be returned if the `Authorization` request header is used.
  - implemented

https://tools.ietf.org/html/rfc6750#section-3.1:

- a 401 SHOULD be returned for an `invalid_token` error
  - implemented
- a 403 SHOULD be returned for an `insufficient_scope` error
  - not implemented, Doorkeeper doesn't seem to use this error currently

toupeira commented on app/controllers/doorkeeper/token_info_controller.rb:
```diff
@@ -6,7 +6,7 @@ def show
         render json: doorkeeper_token, status: :ok
       else
-        error = OAuth::ErrorResponse.new(name: :invalid_request)
+        error = OAuth::InvalidTokenResponse.new
```
I'm not entirely sure if `invalid_token` instead of `invalid_request` is valid here, since the `invalid_token` error is only mentioned in [RFC 6750](https://tools.ietf.org/html/rfc6750#section-3.1).

But it seems appropriate since the code above validates the token, also compare with a similar check in `lib/doorkeeper/rails/helpers.rb`

nbulaj replied:
Checked the document, agree

nbulaj commented:
Hey @toupeira , thanks for this work!

Could you please rebase to current master? SQlite updated this month and there are some issues on CI with it, it currently fixed in latest master.

toupeira commented:
@nbulaj yep encountered the same issue in `doorkeeper-openid_connect` :) Rebased now!

nbulaj commented:
Thanks @toupeira . I'm wondering if we can break somebody project that depends on this status code? :thinking: Maybe we can add some entry to upgrade guides, WDYT ?

nbulaj commented on spec/controllers/authorizations_controller_spec.rb:
```diff
@@ -165,7 +165,7 @@ def translated_error_message(key)
 
     it 'renders 400 error' do
-      expect(response.status).to eq 401
+      expect(response.status).to eq 400
```
:+1: see example description :)

toupeira commented:
> Thanks @toupeira . I'm wondering if we can break somebody project that depends on this status code? :thinking: Maybe we can add some entry to upgrade guides, WDYT ?

@nbulaj yeah I guess this could be considered a breaking API change :) I'd be happy to add something, you mean [this upgrade guide](https://github.com/doorkeeper-gem/doorkeeper/wiki/Migration-from-old-versions) right?

nbulaj commented:
@toupeira yep. Also we can add a note in NEWS.md for those who reads changelogs :)
I'm planning to release it with 5.1, don't think that such change can bring a lot of problems to end users.

nbulaj commented:
Any news here @toupeira ?

toupeira commented:
@nbulaj sorry, dropped off the face of the earth there for a bit ;-)

I added some details to the `NEWS.md` entry now, and also copied it to the wiki.
